### PR TITLE
Base miner training improvements and Content-Aware Model Orchestration (CAMO) Framework 

### PR DESCRIPTION
**Pull request by aliang322**
This PR introduces base miner improvements and a novel multi-model framework for future iterations. Notably,
- New base miners using the [UCF architecture (ICCV 2023)](https://openaccess.thecvf.com/content/ICCV2023/papers/Yan_UCF_Uncovering_Common_Features_for_Generalizable_Deepfake_Detection_ICCV_2023_paper.pdf), the top scoring architecture on the [DeepfakeBench benchmark](https://github.com/SCLBD/DeepfakeBench/tree/main)
  - Training loop and pretrained weight usage for new UCF miner neurons (trained on different data)
  - [Pretrained weights available](https://huggingface.co/bitmind/ucf/tree/main) on our HuggingFace
- Preprocessed dataset creation pipeline for training generalist or expert deepfake detectors
  - Example: A cropped, aligned, and transformed face-only image dataset output readily useable as a [training dataset for deepfake face expert models](https://huggingface.co/collections/bitmind/face-expert-datasets-66cf5ee4a181ad5423f6f477)
- Our most performant base miners yet using a [Content-Aware Model Orchestration (CAMO) Framework for Deepfake Detection](https://www.notion.so/bitmindlabs/CAMO-Content-Aware-Model-Orchestration-CAMO-Framework-for-Deepfake-Detection-43ef46a0f9de403abec7a577a45cd075 )
- Standardized training utils (data preprocessing, loading, splitting) across different architectures/model iterations

### CODE REVIEW

File: base_miner/UCF/README.md [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+This model has been adapted from [DeepfakeBench](https://github.com/SCLBD/DeepfakeBench).
+
+## Setup Instructions
+
+1. **Download Model Weights**:
+   - Download the `ucf_best.pth` and `xception_best.pth` files from the [DeepfakeBench Releases](https://github.com/SCLBD/DeepfakeBench/releases/tag/v1.0.1) page.
+
+2. **Prepare the Weights Directory**:
+   - Create a directory named `weights/` in your project root.
+   - Place the downloaded files into this `weights/` directory.
```
Comment (dylanuys) [inline base_miner/UCF/README.md]: [Suggestion] This happens automatically now right? If so, lets drop the setup instructions

File: base_miner/UCF/config/__init__.py [hunk 2]
```diff
@@ -0,0 +1,7 @@
+import os
+import sys
+current_file_path = os.path.abspath(__file__)
+parent_dir = os.path.dirname(os.path.dirname(current_file_path))
+project_root_dir = os.path.dirname(parent_dir)
+sys.path.append(parent_dir)
+sys.path.append(project_root_dir)
```
Comment (dylanuys) [inline base_miner/UCF/config/__init__.py:7]: [Question] is this necessary when the project is installed in editable mode with `pip install -e .` ?

File: base_miner/UCF/detectors/__init__.py [hunk 3]
```diff
@@ -0,0 +1,11 @@
+import os
+import sys
+current_file_path = os.path.abspath(__file__)
+parent_dir = os.path.dirname(os.path.dirname(current_file_path))
+project_root_dir = os.path.dirname(parent_dir)
+sys.path.append(parent_dir)
+sys.path.append(project_root_dir)
```
Comment (dylanuys) [inline base_miner/UCF/detectors/__init__.py:7]: [Question] same question as with `base_miner/UCF/config/__init__.py`

File: base_miner/UCF/cropped.jpg
Comment (dylanuys) [inline base_miner/UCF/cropped.jpg:1]: [Question] is this needed for testing or should we remove

File: base_miner/UCF/loss/__init__.py [hunk 4]
```diff
@@ -0,0 +1,13 @@
+import os
+import sys
+current_file_path = os.path.abspath(__file__)
+parent_dir = os.path.dirname(os.path.dirname(current_file_path))
+project_root_dir = os.path.dirname(parent_dir)
+sys.path.append(parent_dir)
+sys.path.append(project_root_dir)
```
Comment (dylanuys) [inline base_miner/UCF/loss/__init__.py:7]: [Question] Same question with other __init__ files, I see these come straight from DeepfakeBench, but this is hacky af lol

File: base_miner/UCF/metrics/__init__.py [hunk 5]
```diff
@@ -0,0 +1,7 @@
+import os
+import sys
+current_file_path = os.path.abspath(__file__)
+parent_dir = os.path.dirname(os.path.dirname(current_file_path))
+project_root_dir = os.path.dirname(parent_dir)
+sys.path.append(parent_dir)
+sys.path.append(project_root_dir)
```
Comment (dylanuys) [inline base_miner/UCF/metrics/__init__.py:7]: [Question] same as __init__ files

File: base_miner/UCF/networks/__init__.py [hunk 6]
```diff
@@ -0,0 +1,11 @@
+import os
+import sys
+current_file_path = os.path.abspath(__file__)
+parent_dir = os.path.dirname(os.path.dirname(current_file_path))
+project_root_dir = os.path.dirname(parent_dir)
+sys.path.append(parent_dir)
+sys.path.append(project_root_dir)
```
Comment (dylanuys) [inline base_miner/UCF/networks/__init__.py:7]: [Question] same

File: base_miner/UCF/pretrained_ucf.py [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+from imutils import face_utils
+from skimage import transform as trans
+import gc
+
+from base_miner.UCF.detectors import DETECTOR
+
+class UCF:
+    def __init__(self, config_path="./config/ucf.yaml", weights_dir="./weights/", weights_hf_repo_name="bitmind/ucf",
+                 ucf_checkpoint_name="ucf_best.pth", backbone_checkpoint_name="xception_best.pth",
+                 predictor_path = "../../bitmind/dataset_processing/dlib_tools/shape_predictor_81_face_landmarks.dat"):
```
Comment (dylanuys) [inline base_miner/UCF/pretrained_ucf.py]: [Suggestion] It would be good to have paths like these in a config file (probably somewhere in the UCF dir)

File: base_miner/UCF/sample_images/ffhq_5.jpg
Comment (dylanuys) [inline base_miner/UCF/sample_images/ffhq_5.jpg:1]: [Question] are these sample images used for testing or just adhoc?

File: base_miner/UCF/train_ucf.py [hunk 8]
```diff
@@ -0,0 +1,370 @@
...
+parser.add_argument('--task_target', type=str, default="", help='specify the target of current training task')
+args = parser.parse_args()
+torch.cuda.set_device(args.local_rank)
+print(f"torch.cuda.device(0): {torch.cuda.device(0)}")
+print(f"torch.cuda.get_device_name(0): {torch.cuda.get_device_name(0)}")
+
+def ensure_backbone_is_available(logger,
+                                 weights_dir='./weights/',
+                                 model_filename='xception_best.pth',
+                                 hugging_face_repo_name='bitmind/ucf/'):
```
Comment (dylanuys) [inline base_miner/UCF/train_ucf.py]: [Suggestion] config file for paths would be good, esp since these are shared across different classes

File: base_miner/UCF/trainer/trainer.py [hunk 9]
```diff
@@ -0,0 +1,448 @@
...
+
+# Original: https://github.com/SCLBD/DeepfakeBench/blob/main/training/train.py
+
+import os
+import sys
+current_file_path = os.path.abspath(__file__)
+parent_dir = os.path.dirname(os.path.dirname(current_file_path))
+project_root_dir = os.path.dirname(parent_dir)
+sys.path.append(parent_dir)
+sys.path.append(project_root_dir)
```
Comment (dylanuys) [inline base_miner/UCF/trainer/trainer.py:12]: [Suggestion] To wrap up my 1billion comments on these hacky path management snippets - if removing these is a really annoying lift that requires reformatting all local imports, we can leave it as-is for now and address in a subsequent PR

File: base_miner/object_detection_experiments.ipynb
Comment (dylanuys) [inline base_miner/object_detection_experiments.ipynb:1]: [Suggestion] Lets move this to a new repository (can also just drop it in https://github.com/BitMind-AI/subnet-sandbox until we have a better place)

File: bitmind/real_fake_dataset.py [hunk 10]
```diff
@@ -39,36 +55,46 @@ def __getitem__(self, index: int) -> tuple:
...
+            else:
+                image = self.normalize_image(T.ToTensor()(image))
+                
+        if self.source_label_mapping:
+            source_label = self.source_label_mapping[source.huggingface_dataset_path]
+            return image, label, source_label
+            
         return image, label
-
+
```
Comment (dylanuys) [inline bitmind/real_fake_dataset.py]: [Pickyashell] whitespace

File: bitmind/synthetic_image_generation/combine_datasets.py
Comment (dylanuys) [inline bitmind/synthetic_image_generation/combine_datasets.py:1]: [Suggestion] Flagging this as something that should go in a separate dataset prep repository (that has the subnet repo as a dependency), along with `bitmind/synthetic_data_generation/generate_synthetic_dataset.py` and potentially also `bitmind/dataset_processing/training_dataset_processor.py`

wdyt @benliang99 @aliang322 I chatted w Ken about this thought, reasoning is to keep the subnet repo focused on subnet operations, and have clear separation between this and huggingface dataset prep

File: bitmind/synthetic_image_generation/utils/hugging_face_utils.py
Comment (dylanuys) [inline bitmind/synthetic_image_generation/utils/hugging_face_utils.py:1]: [Suggestion] adding this to the list of things that may belong in a separate repo

File: bitmind/synthetic_image_generation/utils/image_utils.py [hunk 11]
```diff
@@ -1,4 +1,7 @@
 import PIL
+import os
```
Comment (dylanuys) [inline bitmind/synthetic_image_generation/utils/image_utils.py]: [Suggestion] adding this to the list of things that may belong in a separate repo

**Review by dylanuys — CHANGES_REQUESTED** [review]

File: start_miner.sh [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+if pm2 list | grep -q "bitmind_miner"; then
+  echo "Process 'bitmind_miner' is already running. Deleting it..."
+  pm2 delete bitmind_miner
+fi
+
+# Start the process with arguments from environment variables
+pm2 start python --name bitmind_miner -- neurons/miner.py \
+  --neuron.model_path $MODEL_PATH \
+  --netuid $NETUID \
+  --subtensor.network $SUBTENSOR_NETWORK \
```
Comment (elilambnz) [inline start_miner.sh:18]: Can we please add a flag for `--subtensor.chain_endpoint $SUBTENSOR_CHAIN_ENDPOINT \`?

A new env var would need to be added also

File: base_miner/UCF/pretrained_ucf.py (hunk 7 above)
Comment (aliang322) [reply base_miner/UCF/pretrained_ucf.py]: Addressed with new `base_miner/UCF/config/constants.py` file used in `pretrained_ucf.py` UCF class and across miner neurons!

**Review by dylanuys — APPROVED** [review]

File: base_miner/UCF/config/__init__.py (hunk 2 above)
Comment (aliang322) [reply base_miner/UCF/config/__init__.py:7]: Will remove these type of imports and refactor where necessary in a subsequent small PR. Thanks!

File: bitmind/synthetic_image_generation/utils/image_utils.py (hunk 11 above)
Comment (benliang99) [reply bitmind/synthetic_image_generation/utils/image_utils.py]: This one is necessary for image annotation generation, but we may want to reorganize all the utils in the future.

File: bitmind/synthetic_image_generation/utils/hugging_face_utils.py
Comment (benliang99) [reply bitmind/synthetic_image_generation/utils/hugging_face_utils.py:1]: Moved

File: bitmind/synthetic_image_generation/combine_datasets.py
Comment (benliang99) [reply bitmind/synthetic_image_generation/combine_datasets.py:1]: Agreed, moved

File: base_miner/object_detection_experiments.ipynb
Comment (benliang99) [reply base_miner/object_detection_experiments.ipynb:1]: Agreed, moved
